subsys/dfu/src/dfu_target.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include <zephyr.h>
+#include <logging/log.h>
+#include <dfu/mcuboot.h>
+#include <dfu/dfu_target.h>
+#include "dfu_target_mcuboot.h"
+#include "dfu_target_modem.h"
+
+#ifdef CONFIG_DFU_TARGET_MODEM
+extern struct dfu_target dfu_target_modem;
+#endif
+#ifdef CONFIG_DFU_TARGET_MCUBOOT
+extern struct dfu_target dfu_target_mcuboot;
+#endif
+
+#define MIN_SIZE_IDENTIFY_BUF 32
+
+#define MCUBOOT_IMAGE 1
+#define MODEM_DELTA_IMAGE 2
+
+LOG_MODULE_REGISTER(dfu_target, CONFIG_DFU_TARGET_LOG_LEVEL);
+
+static struct dfu_target *ctx;
+static bool   initialized;
+
+int dfu_target_img_type(const void *const buf, size_t len)
+{
+	if (len < MIN_SIZE_IDENTIFY_BUF) {
+		return -EAGAIN;
+	}
+
+	if (IS_ENABLED(CONFIG_BOOTLOADER_MCUBOOT) &&
+	    dfu_target_mcuboot_identify(buf)) {
+		return MCUBOOT_IMAGE;
+	}
+
+	if (IS_ENABLED(CONFIG_DFU_TARGET_MODEM) &&
+	    dfu_target_modem_identify(buf)) {
+		return MODEM_DELTA_IMAGE;
+	}
+
+	LOG_ERR("No supported image type found");
+	return -ENOTSUP;
+}
+
+int dfu_target_init(int img_type)
+{
+	if (initialized) {
+		return 0;
```

Comment:
I'd consider returning `-EBUSY` in this case. If someone tried to run two parallel DFU operations (which is not supported), for instance modem and application, we could get a false impression that `dfu_target_init` succeeded.

Reply:
Good idea :+1: